### DISCUSSION TITLE
Clearer spelling suggestions in CLI

### DISCUSSION THREAD

**Discussion by shannonbux** [gatsbyjs/gatsby#13512] [Ideas / Feature Requests]
## Summary
Suggestion to improve the CLI suggestions when you misspell a command or enter an unsupported/invalid command.

## Details
Currently, the CLI looks like this when you misspell something. 

<img width="569" alt="Screen Shot 2019-04-19 at 6 36 53 PM" src="https://user-images.githubusercontent.com/32467162/56449703-27173f80-62d2-11e9-9fc7-ca9c7b2f9fb9.png">

Here's what happens when you enter an invalid command. Exactly the same as above except the last few lines look like this:
<img width="235" alt="Screen Shot 2019-04-19 at 6 39 38 PM" src="https://user-images.githubusercontent.com/32467162/56449728-81b09b80-62d2-11e9-867d-7d98586ff2de.png">

This is pretty good, and can still be improved because it isn't immediately obvious that 
1. the command didn't work
2. there is a suggestion that the user can act on

## Possible improvements
One design principle is proximity, which means it's helpful to users when the relevant information is close to each other in time and space. So we want these things to be close to each other in time and space:
1. The error explanation
2. The suggestion for what to do next
3. The area where the user can take action and run the new command

If someone accepts an autocorrect suggestion:
```shell
Shannons-MacBook-Pro:watson-dog shannonsoper$ gatsby dvelop
Running `gatsby develop` (press [n] to run `gatsby dvelop` instead)
```

If someone doesn't want the autocorrect behavior:

```shell
Shannons-MacBook-Pro:watson-dog shannonsoper$ gatsby dvelop
Running `gatsby develop` (press n to run `gatsby dvelop` instead) n
Running `gatsby dvelop`
Error: unknown argument: `gatsby dvelop` 
Here are all Gatsby commands. Did you mean one of these?

Usage: gatsby <command> [options]

Commands:
  gatsby develop                   Start development server. Watches files,
                                   rebuilds, and hot reloads if something
                                   changes
  gatsby build                     Build a Gatsby project.
  gatsby serve                     Serve previously built Gatsby site.
  gatsby info                      Get environment information for debugging and
                                   issue reporting
  gatsby clean                     Wipe the local gatsby environment including
                                   built assets and cache
  gatsby repl                      Get a node repl with context of Gatsby
                                   environment, see (add docs link here)
  gatsby new [rootPath] [starter]  Create new Gatsby project.
  gatsby telemetry                 Enable or disable Gatsby anonymous analytics
                                   collection.

Options:
  --verbose                Turn on verbose output     [boolean] [default: false]
  --no-color, --no-colors  Turn off the color in output
                                                      [boolean] [default: false]
  -h, --help               Show help                                   [boolean]
  -v, --version            Show version number                         [boolean]
 
Shannons-MacBook-Pro:watson-dog shannonsoper$
```

OR autocorrect could work great too, sort of like spellchecking in Microsoft Word.

## Motivation
It's good to reduce the amount of cognitive effort people have to spend on small and large things. This is a small thing that we can reduce people's effort to notice and correct. #12951 

## Question
If the CLI talks directly to the user, do we want it to be an "I" or a "we" e.g. "I will run this command for you"? Otherwise, we have to speak in passive voice, e.g. "this command will be run". I'm sort of in favor of "I" with passive voice a close second, though could be convinced otherwise. Not a fan of "we" lol.

Thanks and fyi @DSchau @jamo @KyleAMathews @sidharthachatterjee

**Comment by dev-rsilver** [comment]
I would suggest phrasing as a question directed to the user to avoid use of "I" or "we". "Do you want to run `gatsby develop` [y/n]?"

**Comment by janosh** [comment]
Slightly related: I think the CLI could also be improved by not asking the user if `gatsby-cli` should use a different port if something is already occupying the default one (8000) but _simply do it_. I.e. rather than

```sh
Something is already running at port 8000 
Would you like to run the app at another port instead? [Y/n]
```

just inform the user that the next available port will be used instead:

```sh
Something is already running at port 8000 
Using port 8001 instead.
Rerun `<command name>` with the `-p` option to specify a different port.
```
I made [this suggestion before](https://github.com/gatsbyjs/gatsby/issues/368#issuecomment-413314450) in #368 and would be interested if there's a problem with this approach that escapes me.

**Comment by shannonbux** [comment]
I like both of those ideas!

**Comment by DSchau** [comment]
Hi @shannonbux!

This works!

```
gatsby develo
error Usage: gatsby <command> [options]

Commands:
  gatsby develop                   Start development server. Watches files, rebuilds, and hot reloads if something changes
  gatsby build                     Build a Gatsby project.
  gatsby serve                     Serve previously built Gatsby site.
  gatsby info                      Get environment information for debugging and issue reporting
  gatsby clean                     Wipe the local gatsby environment including built assets and cache
  gatsby repl                      Get a node repl with context of Gatsby environment, see (add docs link here)
  gatsby new [rootPath] [starter]  Create new Gatsby project.
  gatsby plugin                    Useful commands relating to Gatsby plugins
  gatsby telemetry                 Enable or disable Gatsby anonymous analytics collection.

Options:
  --verbose                Turn on verbose output                                                                           [boolean] [default: false]
  --no-color, --no-colors  Turn off the color in output                                                                     [boolean] [default: false]
  -h, --help               Show help                                                                                                         [boolean]
  -v, --version            Show version number                                                                                               [boolean]

Did you mean this?
     develop
```

Maybe we could improve the comprehension of the message so it's clearer?

**Comment by shannonbux** [comment]
That does look like an improvement since it now says the word "error," which wasn't there before.

However "error Usage: gatsby <command> [options] " still isn't quite as helpful as it could be because the user still has to scan the list with no assistance. It could say something like 

```shell
gatsby devlop
error: this command doesn't exist. Did you mean `gatsby develop` or another one of the commands below?

Commands:
  gatsby develop                   Start development server. Watches files, rebuilds, and hot reloads if something changes
  gatsby build                     Build a Gatsby project.
  gatsby serve                     Serve previously built Gatsby site.
  gatsby info                      Get environment information for debugging and issue reporting
  gatsby clean                     Wipe the local gatsby environment including built assets and cache
  gatsby repl                      Get a node repl with context of Gatsby environment, see (add docs link here)
  gatsby new [rootPath] [starter]  Create new Gatsby project.
  gatsby plugin                    Useful commands relating to Gatsby plugins
  gatsby telemetry                 Enable or disable Gatsby anonymous analytics collection.

Options:
  --verbose                Turn on verbose output                                                                           [boolean] [default: false]
  --no-color, --no-colors  Turn off the color in output                                                                     [boolean] [default: false]
  -h, --help               Show help                                                                                                         [boolean]
  -v, --version            Show version number                                                                                               [boolean]
```

or even better, what would it take for us to do what google search does? They just run the thing they think you mean anyway, and tell you they are doing it.
```shell
gatsby devlop
running `gatsby develop` now. Press [n] to stop
```

**Comment by shannonbux** [comment]
Figma file ready for commenting: https://www.figma.com/file/z81zOLund5Z5FCjnHnUIrN/CLI?node-id=195%3A40
